libs/wingc/src/type_check.rs

```diff
@@ -2922,4 +2951,126 @@ mod tests {
+
+		// T is a subtype of T? since T can be used anywhere a T? is expected
+		// (but not vice versa)
+		assert!(string.is_subtype_of(&opt_string));
+		assert!(!opt_string.is_subtype_of(&string));
+
+		// subtyping is reflexive
+		assert!(string.is_subtype_of(&string));
+		assert!(opt_string.is_subtype_of(&opt_string));
+	}
+
+	#[test]
+	fn function_subtyping_across_phases() {
+		let void = UnsafeRef::<Type>(&Type::Void as *const Type);
+		let inflight_fn = Type::Function(FunctionSignature {
+			parameters: vec![],
+			return_type: void,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
+		let preflight_fn = Type::Function(FunctionSignature {
+			parameters: vec![],
+			return_type: void,
+			phase: Phase::Preflight,
+			js_override: None,
+		});
+
+		// functions of different phases are not subtypes of each other
+		assert!(!inflight_fn.is_subtype_of(&preflight_fn));
+		assert!(!preflight_fn.is_subtype_of(&inflight_fn));
+
+		// subtyping is reflexive
+		assert!(inflight_fn.is_subtype_of(&inflight_fn));
+		assert!(preflight_fn.is_subtype_of(&preflight_fn));
+	}
+
+	#[test]
+	fn function_subtyping_incompatible_single_param() {
+		let void = UnsafeRef::<Type>(&Type::Void as *const Type);
+		let num = UnsafeRef::<Type>(&Type::Number as *const Type);
+		let string = UnsafeRef::<Type>(&Type::String as *const Type);
+		let num_fn = Type::Function(FunctionSignature {
+			parameters: vec![num],
+			return_type: void,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
+		let str_fn = Type::Function(FunctionSignature {
+			parameters: vec![string],
+			return_type: void,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
+
+		// functions of incompatible arguments are not subtypes of each other
+		assert!(!num_fn.is_subtype_of(&str_fn));
+		assert!(!str_fn.is_subtype_of(&num_fn));
+	}
+
+	#[test]
+	fn function_subtyping_incompatible_return_type() {
+		let void = UnsafeRef::<Type>(&Type::Void as *const Type);
+		let num = UnsafeRef::<Type>(&Type::Number as *const Type);
+		let string = UnsafeRef::<Type>(&Type::String as *const Type);
+		let returns_num = Type::Function(FunctionSignature {
+			parameters: vec![],
+			return_type: num,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
+		let returns_str = Type::Function(FunctionSignature {
+			parameters: vec![],
+			return_type: string,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
+		let returns_void = Type::Function(FunctionSignature {
+			parameters: vec![],
+			return_type: void,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
+
+		// functions of incompatible return types are not subtypes of each other
+		assert!(!returns_num.is_subtype_of(&returns_str));
+		assert!(!returns_str.is_subtype_of(&returns_num));
+
+		// functions with specific return types are subtypes of functions with void return type
+		assert!(returns_num.is_subtype_of(&returns_void));
+		assert!(returns_str.is_subtype_of(&returns_void));
```

Comment:
I'm not sure I approve of this relationship needed :)  Why is this needed?

Reply:
The reason I have this in place is so that both of these kinds of existing usages in tests will work:

```js
let fn1 = new cloud.Function(inflight (m: str) {
  print("im not returning anything");
});

let fn2 = new cloud.Function(inflight (m: str): str {
  return "Hello, ${m}!";
});

let usage = inflight () => {
  fn1.invoke("");
  let result = fn2.invoke("");
};
```

We don't have generics, so we're kind of in a tough place as far as offering flexible parameter and return types, since `fn.invoke()` needs to always return the SAME type.

The alternative to this (if we don't want this special return-type subtyping rule) is that every `cloud.Function` must have the same return type, which would have to be `str` or `Json`. We could change it to `str` or `Json` but it would be a breaking change at the moment.

Let's open a separate issue about this because I don't think it's core to this PR

Reply:
Related: https://github.com/winglang/wing/issues/1843